README.md
 Nigerian Phone Number Validator (PHP)
[![HitCount](http://hits.dwyl.io/djunehor/php_validate_nigerian_phone.svg)](http://hits.dwyl.io/djunehor/php_validate_nigerian_phone) [![CircleCI](https://circleci.com/gh/djunehor/php_validate_nigerian_phone.svg?style=svg)](https://circleci.com/gh/djunehor/php_validate_nigerian_phone)

#### Issues and pull requests welcome.

A PHP module to validate and format a Nigerian phone number as well as deduce the network provider or area code.

## NOTE: The network resolution function can't be accurate because Nigeria implemented [Mobile Number Portability](https://en.wikipedia.org/wiki/Mobile_number_portability) in 2013, so the number prefix cannot be reliably used to determine operator anymore.

### Table of Contents
* [Installation](#installation)
* [Usage](#usage)
* [Features](#features)
* [Contribute](#contribute)
* [Run Tests](#tests)

## Installation
You will need >[PHP 5.6](https://www.php.net/) and [composer](https://getcomposer.org/download/).

Install using composer: `composer require djunehor/validate_nigerian_phone`

### In a Laravel project
* If using Laravel 5.5 and above, no further action required
* If you're using Laravel < 5.5, you'll need to register the service provider. Open up config/app.php and add the following to the providers array:
`Djunehor\Validator\PhoneValidatorServiceProvider::class`

## Usage
### In Laravel
```php
...

/*This checks if the number supplied
*is a valid nigerian number
* either land line or mobile number
*/
public function controllerName(Request $request) {
		$request->validate( [
			'phone' => 'required|ngphone'
		]);
	}
	
//to validate if it's a mobile number
$request->validate( [
			'phone' => 'required|ngphone:mobile'
		]);
		
//to validate if it's a landline
$request->validate( [
			'phone' => 'required|ngphone:land'
		]);		
```

# In a PHP project
If autoload is not enabled, add `require vendor/autoload.php` to the top of the file, then
```php
use \Djunehor\Validator\NigerianPhone;

$phone = new NigerianPhone('+2348135087966');

// Check if is valid
$phone->isValid(); // true

// Get formatted
$phone->formatted(); // 08135087966

// Get Network
$phone->getNetwork(); // mtn

// Check if is mtn
$phone->isMtn(); // True


// Get network from phone number prefix e.g
$phone->getNetworkByPrefix('0703'); // mtn

```

## Features
### Currently implemented
* isValid
* formatted
* getNetwork
* getAreaCode
* isMtn
* isGlo
* isAirtel
* is9mobile
* isSmile
* isMultilinks
* isVisafone
* isNtel
* isStarcomms
* isZoom
* getPrefixesByNetwork
* getNetworkByPrefix
* getAreaCodeByName

### Tests
* Run `phpunit tests/NigerianPhoneTest.php`

## Contribute
Check out the issues on GitHub and/or make a pull request to contribute!


CHANGELOG.md
# Changelog

All notable changes to `validate_nigerian_phone` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.



tests/NigerianPhoneTest.php
<?php

declare(strict_types=1);

use \Djunehor\Validator\NigerianPhone;
use \PHPUnit\Framework\TestCase;

class NigerianPhoneTest extends TestCase {

	public static $validNumbers = [
		'+2348135087966',
		'2348135087966',
		'08135087966',
		'8135087966',
		'04221542',
//		'+234 0 256 7854',
		'0425 5832'
	];

	public static $phones = [
		'+2348135087966' => 'mtn',
		'08022334455'    => 'airtel',
		'2348182334455' => '9mobile',
		'8052233455'    => 'glo',
		'08197456754'    => 'starcomms',
//		'08117456754'    => null,
	];

	public static $wrongPhones = [
		'+2346135087966',
		'06022334455',
	];

	public static $landLines = [
		'04221542'        => 'Enugu',
//		'+234 5 2569 754' => 'Benin City',
		'0355 5832'       => 'Oshogbo'
	];

	public static $wrongLandLines = [
		'00221542',
		'+234 9 256 7854',
		'0995 5832'
	];

	public static $wrongNumbers = [
		'2345067546789',
		'90239289'
	];

	public static $correctPrefixes = [
		'0703' => 'mtn',
		'0802' => 'airtel',
		'0809' => '9mobile',
		'0707' => 'zoom'
	];

	public static $wrongPrefixes = [
		'0801' => null,
		'0602' => null,
		'0819' => null,
	];

	public static $areaCodes = [
		'Ilorin'        => '31',
		'New Bussa'     => '33',
		'Akure'         => '34',
		'Munami' => null
	];

	public function testValidNumbers() {
		foreach ( self::$validNumbers as $case ) {
			$app = new NigerianPhone( $case );

			$this->assertTrue( $app->isValid() );
		}
	}

	public function testInvalidNumbers() {
		foreach ( self::$wrongNumbers as $case ) {
			$app = new NigerianPhone( $case );
			$this->assertTrue( $app->isValid() == false );
		}
	}

	public function testIsPhone() {
		foreach ( self::$phones as $phone => $value ) {
			$app = new NigerianPhone( $phone );

			$this->assertTrue( $app->isValidPhone );
		}
	}

	public function testIsNotPhone() {
		foreach ( self::$wrongPhones as $phone => $value ) {
			$app = new NigerianPhone( $phone );
			$this->assertTrue( $app->isValidPhone == false );
		}
	}

	public function testIsLandLine() {
		foreach ( self::$landLines as $phone => $value ) {
			$app = new NigerianPhone( $phone );
			$this->assertTrue( $app->isValidLandLine );
		}
	}

	public function testIsNotLandLine() {
		foreach ( self::$wrongLandLines as $phone => $value ) {
			$app = new NigerianPhone( $phone );
			$this->assertFalse( $app->isValidLandLine );
		}
	}

	public function testNetwork() {
		foreach ( self::$phones as $phone => $network ) {
			$app = new NigerianPhone( $phone );
			$this->assertTrue( $app->getNetwork() == $network );
		}
	}

	public function testAreaCode() {
		foreach ( self::$landLines as $phone => $network ) {
			$app = new NigerianPhone( $phone );
			$this->assertTrue( $app->getAreaCode() == $network );
		}
	}

	public function testPrefix() {
		foreach ( self::$correctPrefixes as $prefix => $network ) {
			$app = new NigerianPhone();
			$this->assertTrue( $app->getNetworkByPrefix( $prefix ) == $network );
		}
	}

	public function testGetAreaCodeByName() {
		foreach ( self::$areaCodes as $area => $code ) {
			$app = new NigerianPhone();
			$this->assertTrue( $app->getAreaCodeByName( $area ) == $code );
		}
	}
}


src/PhoneValidatorServiceProvider.php
<?php

namespace Djunehor\Validator;
use Validator;

use Illuminate\Support\ServiceProvider;
use Djunehor\Validator\NigerianPhone;

class PhoneValidatorServiceProvider extends ServiceProvider {
	/**
	 * Bootstrap the application services.
	 *
	 * @return void
	 */

	protected $message = 'The phone number you provided is not valid!';

	public function boot() {
		Validator::extend('ngphone', function ($attribute, $value, $parameters, $validator) {

			$phone = new NigerianPhone($value);
			$type = (!empty($parameters)) ? head($parameters) : null;

			switch ($type) {
				case 'mobile':
					return $phone->isValidPhone;
					break;
				case 'land':
					return $phone->isValidLandLine;

				default:
					return $phone->isValid();
			}

		}, $this->message);
	}

	/**
	 * Register the application services.
	 *
	 * @return void
	 */
	public function register() {

	}

}


src/NigerianPhone.php
<?php
/*
Package Name: Nigerian Phone Number Validator
Description: Class to validate a Nigerian phone number as well as attempt to deduce the network.
Version: 1.0
Author: Zacchaeus (Djunehor) Bolaji
Author URI: https://github.com/djunehor
*/
namespace Djunehor\Validator;

class NigerianPhone {

	public static $allPrefixes = [
		'mtn'        => [
			'0803',
			'0703',
			'0903',
			'0806',
			'0706',
			'0813',
			'0814',
			'0816',
			'0810',
			'0906',
			'07025',
			'07026',
			'0704'
		],
		'glo'        => [ '0805', '0705', '0905', '0807', '0815', '0905', '0811' ],
		'airtel'     => [ '0802', '0902', '0701', '0808', '0708', '0812', '0901', '0907' ],
		'9mobile'    => [ '0809', '0909', '0817', '0818', '0908' ],
		'ntel'       => [ '0804' ],
		'smile'      => [ '0702' ],
		'multilinks' => [ '0709', '07027' ],
		'visafone'   => [ '07025', '07026', '0704' ],
		'starcomms'  => [ '07028', '07029', '0819' ],
		'zoom'       => [ '0707' ],
	];

	public static $allAreaCodes = [
		'Lagos'         => '01',
		'Ibadan'        => '02',
		'Abuja'         => '09',
		'Ado-Ekiti'     => '30',
		'Ilorin'        => '31',
		'New Bussa'     => '33',
		'Akure'         => '34',
		'Oshogbo'       => '35',
		'Ile-Ife'       => '36',
		'Ijebu-Ode'     => '37',
		'Oyo'           => '38',
		'Abeokuta'      => '39',
		'Wukari'        => '41',
		'Enugu'         => '42',
		'Abakaliki'     => '43',
		'Makurdi'       => '44',
		'Ogoja'         => '45',
		'Onitsha'       => '46',
		'Lafia'         => '47',
		'Awka'          => '48',
		'Ikare'         => '50',
		'Owo'           => '51',
		'Benin City'    => '52',
		'Warri'         => '53',
		'Sapele'        => '54',
		'Agbor'         => '55',
		'Asaba'         => '56',
		'Auchi'         => '57',
		'Lokoja'        => '58',
		'Okitipupa'     => '59',
		'Sokoto'        => '60',
		'Kafanchan'     => '61',
		'Kaduna'        => '62',
		'Gusau'         => '63',
		'Kano'          => '64',
		'Katsina'       => '65',
		'Minna'         => '66',
		'Kontagora'     => '67',
		'Birnin-Kebbi'  => '68',
		'Zaria'         => '69',
		'Pankshin'      => '73',
		'Azare'         => '71',
		'Gombe'         => '72',
		'Jos'           => '73',
		'Yola'          => '75',
		'Maiduguri'     => '76',
		'Bauchi'        => '77',
		'Hadejia'       => '78',
		'Jalingo'       => '79',
		'Aba, Nigeria'  => '82',
		'Owerri'        => '83',
		'Port Harcourt' => '84',
		'Uyo'           => '85',
		'Ahoada'        => '86',
		'Calabar'       => '87',
		'Umuahia'       => '88',
		'Yenagoa'       => '89',
		'Ubiaja'        => '55',
		'Kwara'         => '31',
		'Igarra'        => '57',
		'Ughelli'       => '53',
		'Uromi'         => '57'
	];

	public $prefixes, $areaCodes;
	public $isValidPhone = false;
	public $isValidLandLine = false;
	public $line = '';

	public function __construct( $line = null ) {
		foreach ( self::$allPrefixes as $key => $value ) {
			foreach ( $value as $item ) {
				$this->prefixes[] = $item;
			}
		}

		foreach ( self::$allAreaCodes as $key => $value ) {
			$this->areaCodes[] = $value;
		}

		if ($line) {
			$this->line = preg_replace( "/[^0-9]/", "", $line );
			$this->isValid();
		}

	}

	public function setLine($line) {
		$this->line = $line;
	}

	public function isValid() {
		# contains country calling code
		if ( substr( $this->line, 0, 3 ) == '234' ) {
			# phone number
			if ( strlen( $this->line ) == 13 ) {
				# remove country code for further analysis
				$stripCcc = str_replace( substr( $this->line, 0, 3 ), "0", $this->line );

				# contains valid phone prefix
				if ( in_array( substr( $stripCcc, 0, 5 ), $this->prefixes ) || in_array( substr( $stripCcc, 0, 4 ), $this->prefixes ) ) {
					$this->isValidPhone = true;

					return true;
				}
			}

			# is a land line
			elseif ( strlen( ( $this->line ) == 11 ) ) {
				# replace country code (234)
				$stripCcc = str_replace( substr( $this->line, 0, 3 ), "", $this->line );

				# contains valid 2-digit area code
				if ( in_array( substr( $stripCcc, 0, 2 ), $this->areaCodes )
				     && strlen( $stripCcc ) == 7) {
					$this->isValidLandLine = true;

					return true;
				}
			}
		}
		# doesn't contain country calling code
		else {
			# check if it starts with any prefix [0807,0906 etc..]
			if ( substr( $this->line, 0, 1 ) == '0' && strlen( $this->line ) == 11 && ( in_array( substr( $this->line, 0, 5 ), $this->prefixes ) || in_array( substr( $this->line, 0, 4 ), $this->prefixes ) ) ) {
				$this->isValidPhone = true;

				return true;
			}
			# check if it starts with any prefix without 0 e.g [807,906, 701 etc..]
			elseif ( strlen( $this->line ) == 10 && ( in_array( substr( $this->line, 0, 4 ), $this->prefixes ) || in_array( substr( $this->line, 0, 3 ), $this->prefixes ) ) ) {
				# add the missing zero for completion
				$this->line         = '0' . $this->line;
				$this->isValidPhone = true;

				return true;
			}
			# check if it's a land line starting with 0
			elseif ( substr( $this->line, 0, 1 ) == 0
			         && in_array( str_replace( substr( $this->line, 0, 1 ), "", substr( $this->line, 0, 3 ) ), $this->areaCodes )
			         && strlen( $this->line ) == 8 ) {
				$this->isValidLandLine = true;

				return true;
			}
		}

		return false;
	}

	public function formatted() {
		if ( substr( $this->line, 0, 3 ) == '234' ) {
			return str_replace( substr( $this->line, 0, 3 ), '0', $this->line );
		}

		return $this->line;
	}

	public function getNetwork() {
		if ( $this->isMtn() ) {
			return 'mtn';
		}
		elseif ( $this->isGlo() ) {
			return 'glo';
		}
		elseif ( $this->isAirtel() ) {
			return 'airtel';
		}
		elseif ( $this->is9mobile() ) {
			return '9mobile';
		}
		elseif ( $this->isSmile() ) {
			return 'smile';
		}
		elseif ( $this->isMultilinks() ) {
			return 'multilinks';
		}
		elseif ( $this->isVisafone() ) {
			return 'visafone';
		}
		elseif ( $this->isNtel() ) {
			return 'ntel';
		}
		elseif ( $this->isStarcomms() ) {
			return 'starcomms';
		}
		elseif ( $this->isZoom() ) {
			return 'zoom';
		}
		else {
			return 'unknown';
		}
	}

	public function getAreaCode() {
		$formatted   = $this->formatted();
		$removedZero = str_replace( '0', '', $formatted );
		$areaCode        = substr( $removedZero, 0, 2 );
		foreach ( self::$allAreaCodes as $area => $code ) {
			if ( $areaCode == $code ) {
				return $area;
			}

		}
		return null;
	}

	public function isMtn() {
		return ( in_array( substr( $this->formatted(), 0, 5 ), self::$allPrefixes['visafone'] ) || in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['mtn'] ) );
	}

	public function isGlo() {
		return ( in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['glo'] ) );
	}

	public function isAirtel() {
		return ( in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['airtel'] ) );
	}

	public function is9mobile() {
		return ( in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['9mobile'] ) );
	}

	public function isSmile() {
		return ( in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['smile'] ) );
	}

	public function isMultilinks() {
		return ( in_array( substr( $this->formatted(), 0, 5 ), self::$allPrefixes['multilinks'] )
		         || in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['multilinks'] ) );
	}

	public function isVisafone() {
		return ( in_array( substr( $this->formatted(), 0, 5 ), self::$allPrefixes['visafone'] )
		         || in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['visafone'] ) );
	}

	public function isNtel() {
		return ( in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['multilinks'] ) );
	}

	public function isStarcomms() {
		return ( in_array( substr( $this->formatted(), 0, 5 ), self::$allPrefixes['starcomms'] )
		         || in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['starcomms'] ) );
	}

	public function isZoom() {
		return ( in_array( substr( $this->formatted(), 0, 4 ), self::$allPrefixes['zoom'] ) );
	}

	public function getPrefixesByNetwork( $network ) {
		if ( array_key_exists( $network, self::$allPrefixes ) ) {
			return self::$allPrefixes[ $network ];
		}

		return [];
	}

	public function getNetworkByPrefix( $area ) {
		foreach (self::$allPrefixes as $key => $value ) {
			if (in_array($area, $value))
				return $key;
		}

		return null;
	}

	public function getAreaCodeByName( $area ) {
		foreach ( self::$allAreaCodes as $key => $value ) {
			if ( $key == $area ) {
				return $value;
			}
		}

		return null;
	}
}

